# Chapter 2
# Data Munging

# Introduction

It is said that around 50% of the data scientist's time goes into transforming raw data into a usable format. Raw data can be in any format or size. It can be structured like RDBMS, semi-structured like CSV, or unstructured like regular text files. These contain some valuable information. And to extract that information, it has to be converted into a data structure or a usable format from which an algorithm can find valuable insights. 

This chapter will guide you through data munging, or the process of preparing the data. It covers the following topics:


- What is data munging?
- DataFrames.jl
- Uploading data from a file 
- Finding the required data 
- Joins and indexing 
- Split-Apply-Combine strategy
- Reshaping the data
- Formula (ModelFrame and ModelMatrix) 
- PooledDataArray
- Web scraping

# 1. What is data munging?

It is considered one of the most essential parts of the data science process; it involves collecting, aggregating, cleaning, and organizing the data to be consumed by the algorithms designed to make discoveries or to create models. This involves numerous steps, including extracting data from the data source and then parsing or transforming the data into a predefined data structure. Data munging is also referred to as data wrangling.

# The data munging process

So what's the data munging process? As mentioned, data can be in any format and the data science process may require data from multiple sources. This data aggregation phase includes scraping it from websites, downloading thousands of .txt or .log files, or gathering the data from RDBMS or NoSQL data stores.

It is very rare to find data in a format that can be used directly by the data science process. The data received is generally in a format unsuitable for modeling and analysis. Generally, algorithms require data to be stored in a tabular format or in matrices. This phase of converting the gathered raw data into the required format can get very complex and time consuming. But this phase creates the foundation of the sophisticated data analysis that can now be done.

It is good to define the structure of the data that you will be feeding the algorithms in advance. This data structure is defined according to the nature of the problem. The algorithms that you have designed or will be designing should not just be able to accept this format of data, but they should also be able to easily identify the patterns, find the outliers, make discoveries, or meet whatever the desired outcomes are.

After defining how the data will be structured, you define the process to achieve that. This is like a pipeline that will accept some forms of data and will give out meaningful data in a predefined format. This phase consists of various steps. These steps include converting data from one form to another, which may or may not require string operations or regular expressions, and finding the missing values and outliers.
                                                                                                
Generally, data science problems revolve around two kinds of data. These two kinds of data will be either categorical or numerical. 

1. Categorical data comes with labels. These labels are formed by some group of values. For example, we can treat weather with categorical features.

2. Numerical data is much more common, for example, temperature. Temperature will be in floating-point numbers and we can certainly apply mathematical operations on it. 

# 2. What is a DataFrame?

A DataFrame is a data structure that has labeled columns, which individually may have different data types. Like a SQL table or a spreadsheet, it has two dimensions. 

DataFrames are the recommended data structure for statistical analysis. Julia provides a package called DataFrames.jl , which have all necessary functions to work with DataFrames.


Julia's package, DataFrames, provides three data types:
- NA: A missing value in Julia is represented by a specific data type, NA.
- DataArray: The array type defined in the standard Julia library, though it has many features, doesn't provide any specific functionalities for data analysis. DataArray provided in DataFrames.jl provides such features (for example, if we required to store in an array some missing values).
- DataFrame: DataFrame is 2-D data structure, like spreadsheets. It is much like R or pandas's DataFrames, and provides many functionalities to represent and analyze data.

# The NA data type and its importance

The DataFrames package brings with it DataArray packages, which provide NA data type. Multiple dispatch is one of the most powerful features of Julia and NA is one such example. Julia has NA type, which provides the singleton object NA that we are using to represent missing values.

In [1]:
x = [1.1, 2.2, 3.3, 4.4, 5.5, 6.6]

6-element Array{Float64,1}:
 1.1
 2.2
 3.3
 4.4
 5.5
 6.6

In [2]:
x[1]=NA

UndefVarError: UndefVarError: NA not defined

Therefore, right now we cannot add NA values to the array that we have created.

So, to load the data into an array that does have NA values, we use DataArray. This
enables us to have NA values in our dataset:


In [3]:
# using DataArrays
# x = DataArray([1.1, 2.2, 3.3, 4.4, 5.5, 6.6])

Therefore, by using DataArrays, we can handle missing data. One more feature provided is that NA doesn't always affect functions applied on the particular dataset. So, the method that doesn't involve an NA value or is not affected by it can be applied on the dataset. If it does involve the NA value, then it will give NA as the result.


# DataArray – a series-like data structure

There are other features similar to Julia's Array type. Type aliases of Vector (one- dimensional Array type) and Matrix (two-dimensional Array type) are DataVector and DataMatrix provided by DataArray.


Creating a 1-D DataArray is similar to creating an Array:


In [4]:
# using DataArrays
# dvector = data([10,20,30,40,50])

In [5]:
Array{Float64,1}([10,20,30,40,50],)

5-element Array{Float64,1}:
 10.0
 20.0
 30.0
 40.0
 50.0

Here, we have NA values, unlike in Arrays. Similarly, we can create a 2-D DataArray, which will be a DataMatrix:

In [6]:
Array{Float64,2}([10 20 30 ; 40 50 60],)

2×3 Array{Float64,2}:
 10.0  20.0  30.0
 40.0  50.0  60.0

Note: DataArrays PACKAGE IS DEPRECATED with Julia versions above 0.7. Use Array{Union{T, Missing}}

https://github.com/JuliaStats/DataArrays.jl

# DataFrames – tabular data structures

Arguably, this is the most important and commonly used data type in statistical computing, whether it is in R (data.frame) or Python (Pandas). This is due to the fact that all the real- world data is mostly in tabular or spreadsheet-like format. This cannot be represented by a simple DataArray:

In [7]:
using DataFrames

In [8]:
df = DataFrame(Name = ["Ajava Rhodiumhi", "Las Hushjoin"], Count = [14.04, 17.3],OS = ["Ubuntu", "Mint"])
 

2×3 DataFrame
│ Row │ Name            │ Count   │ OS     │
│     │ String          │ Float64 │ String │
├─────┼─────────────────┼─────────┼────────┤
│ 1   │ Ajava Rhodiumhi │ 14.04   │ Ubuntu │
│ 2   │ Las Hushjoin    │ 17.3    │ Mint   │

Using the keyword arguments, column names can be defined.

Let's take another example by constructing a new DataFrame:

In [9]:
df2 = DataFrame()

0×0 DataFrame

In [10]:
df2[:X] = 1:10


┌ Warning: `setindex!(df::DataFrame, v::AbstractVector, col_ind::ColumnIndex)` is deprecated, use `begin
│     df[!, col_ind] = v
│     df
│ end` instead.
│   caller = top-level scope at In[10]:1
└ @ Core In[10]:1


1:10

In [11]:
df2[:Y] = ["Head", "Tail","Head", "Head","Tail", "Head","Tail", "Tail","Head", "Tail"]

┌ Warning: `setindex!(df::DataFrame, v::AbstractVector, col_ind::ColumnIndex)` is deprecated, use `begin
│     df[!, col_ind] = v
│     df
│ end` instead.
│   caller = top-level scope at In[11]:1
└ @ Core In[11]:1


10-element Array{String,1}:
 "Head"
 "Tail"
 "Head"
 "Head"
 "Tail"
 "Head"
 "Tail"
 "Tail"
 "Head"
 "Tail"

In [12]:
df2

10×2 DataFrame
│ Row │ X     │ Y      │
│     │ Int64 │ String │
├─────┼───────┼────────┤
│ 1   │ 1     │ Head   │
│ 2   │ 2     │ Tail   │
│ 3   │ 3     │ Head   │
│ 4   │ 4     │ Head   │
│ 5   │ 5     │ Tail   │
│ 6   │ 6     │ Head   │
│ 7   │ 7     │ Tail   │
│ 8   │ 8     │ Tail   │
│ 9   │ 9     │ Head   │
│ 10  │ 10    │ Tail   │

To find out the size of the DataFrame created, we use the size function:

In [13]:
size(df2)

(10, 2)

To view the first few lines of the dataset, we use head(), and for the last few lines, we use
the tail() function:

In [14]:
first(df2,7)

7×2 DataFrame
│ Row │ X     │ Y      │
│     │ Int64 │ String │
├─────┼───────┼────────┤
│ 1   │ 1     │ Head   │
│ 2   │ 2     │ Tail   │
│ 3   │ 3     │ Head   │
│ 4   │ 4     │ Head   │
│ 5   │ 5     │ Tail   │
│ 6   │ 6     │ Head   │
│ 7   │ 7     │ Tail   │

As we have given names to the columns of the DataFrame, these can be accessed using these names.

In [15]:
df2

10×2 DataFrame
│ Row │ X     │ Y      │
│     │ Int64 │ String │
├─────┼───────┼────────┤
│ 1   │ 1     │ Head   │
│ 2   │ 2     │ Tail   │
│ 3   │ 3     │ Head   │
│ 4   │ 4     │ Head   │
│ 5   │ 5     │ Tail   │
│ 6   │ 6     │ Head   │
│ 7   │ 7     │ Tail   │
│ 8   │ 8     │ Tail   │
│ 9   │ 9     │ Head   │
│ 10  │ 10    │ Tail   │

If needed, we can also rename using these columns by using the rename function:

In [16]:
rename!(df2, Symbol("X") => :NewX)

10×2 DataFrame
│ Row │ NewX  │ Y      │
│     │ Int64 │ String │
├─────┼───────┼────────┤
│ 1   │ 1     │ Head   │
│ 2   │ 2     │ Tail   │
│ 3   │ 3     │ Head   │
│ 4   │ 4     │ Head   │
│ 5   │ 5     │ Tail   │
│ 6   │ 6     │ Head   │
│ 7   │ 7     │ Tail   │
│ 8   │ 8     │ Tail   │
│ 9   │ 9     │ Head   │
│ 10  │ 10    │ Tail   │

If there is a need to rename multiple columns, then it is done by using this:

In [17]:
rename!(df2, Symbol("NewX") => :newX, Symbol("Y") => :newY)

10×2 DataFrame
│ Row │ newX  │ newY   │
│     │ Int64 │ String │
├─────┼───────┼────────┤
│ 1   │ 1     │ Head   │
│ 2   │ 2     │ Tail   │
│ 3   │ 3     │ Head   │
│ 4   │ 4     │ Head   │
│ 5   │ 5     │ Tail   │
│ 6   │ 6     │ Head   │
│ 7   │ 7     │ Tail   │
│ 8   │ 8     │ Tail   │
│ 9   │ 9     │ Head   │
│ 10  │ 10    │ Tail   │

Julia also provides a function called describe(), which summarizes the entire dataset. For a dataset with many columns, it can turn out to be very useful:

In [18]:
describe(df2) 

2×8 DataFrame. Omitted printing of 1 columns
│ Row │ variable │ mean   │ min  │ median │ max  │ nunique │ nmissing │
│     │ Symbol   │ Union… │ Any  │ Union… │ Any  │ Union…  │ Nothing  │
├─────┼──────────┼────────┼──────┼────────┼──────┼─────────┼──────────┤
│ 1   │ newX     │ 5.5    │ 1    │ 5.5    │ 10   │         │          │
│ 2   │ newY     │        │ Head │        │ Tail │ 2       │          │

# Installation and using DataFrames.jl

Installation is quite straightforward as it is a registered Julia package:


In [19]:
Pkg.update()
Pkg.add("DataFrames")


UndefVarError: UndefVarError: Pkg not defined

This adds all the required packages to the current namespace. To use the DataFrames package:


In [20]:
using DataFrames

It is also good to have classical datasets that are common for learning purposes. These datasets can be found in the RDatasets package:


In [21]:
Pkg.add("RDatasets")

UndefVarError: UndefVarError: Pkg not defined

In [22]:
using RDatasets

The list of the R packages available can be found using:


In [23]:
iris_dataset = dataset("datasets", "iris")

150×5 DataFrame
│ Row │ SepalLength │ SepalWidth │ PetalLength │ PetalWidth │ Species   │
│     │ Float64     │ Float64    │ Float64     │ Float64    │ Cat…      │
├─────┼─────────────┼────────────┼─────────────┼────────────┼───────────┤
│ 1   │ 5.1         │ 3.5        │ 1.4         │ 0.2        │ setosa    │
│ 2   │ 4.9         │ 3.0        │ 1.4         │ 0.2        │ setosa    │
│ 3   │ 4.7         │ 3.2        │ 1.3         │ 0.2        │ setosa    │
│ 4   │ 4.6         │ 3.1        │ 1.5         │ 0.2        │ setosa    │
│ 5   │ 5.0         │ 3.6        │ 1.4         │ 0.2        │ setosa    │
│ 6   │ 5.4         │ 3.9        │ 1.7         │ 0.4        │ setosa    │
│ 7   │ 4.6         │ 3.4        │ 1.4         │ 0.3        │ setosa    │
│ 8   │ 5.0         │ 3.4        │ 1.5         │ 0.2        │ setosa    │
│ 9   │ 4.4         │ 2.9        │ 1.4         │ 0.2        │ setosa    │
│ 10  │ 4.9         │ 3.1        │ 1.5         │ 0.1        │ setosa    │
⋮
│ 140 │ 6.9         │ 3.1        │ 5.4         │ 2.1        │ virginica │
│ 141 │ 6.7         │ 3.1        │ 5.6         │ 2.4        │ virginica │
│ 142 │ 6.9         │ 3.1        │ 5.1         │ 2.3        │ virginica │
│ 143 │ 5.8         │ 2.7        │ 5.1         │ 1.9        │ virginica │
│ 144 │ 6.8         │ 3.2        │ 5.9         │ 2.3        │ virginica │
│ 145 │ 6.7         │ 3.3        │ 5.7         │ 2.5        │ virginica │
│ 146 │ 6.7         │ 3.0        │ 5.2         │ 2.3        │ virginica │
│ 147 │ 6.3         │ 2.5        │ 5.0         │ 1.9        │ virginica │
│ 148 │ 6.5         │ 3.0        │ 5.2         │ 2.0        │ virginica │
│ 149 │ 6.2         │ 3.4        │ 5.4         │ 2.3        │ virginica │
│ 150 │ 5.9         │ 3.0        │ 5.1         │ 1.8        │ virginica │

You can see that the dataset() function has returned a DataFrame. The dataset contains five columns: SepalLength, SepalWidth, PetalLength, PetalWidth, and Species. It is quite easy to understand the data. A large number of samples have been taken for every species, and the length and width of sepal and petal have been measured, which can be used later to distinguish between them:

In [24]:
first(iris_dataset,10)

10×5 DataFrame
│ Row │ SepalLength │ SepalWidth │ PetalLength │ PetalWidth │ Species │
│     │ Float64     │ Float64    │ Float64     │ Float64    │ Cat…    │
├─────┼─────────────┼────────────┼─────────────┼────────────┼─────────┤
│ 1   │ 5.1         │ 3.5        │ 1.4         │ 0.2        │ setosa  │
│ 2   │ 4.9         │ 3.0        │ 1.4         │ 0.2        │ setosa  │
│ 3   │ 4.7         │ 3.2        │ 1.3         │ 0.2        │ setosa  │
│ 4   │ 4.6         │ 3.1        │ 1.5         │ 0.2        │ setosa  │
│ 5   │ 5.0         │ 3.6        │ 1.4         │ 0.2        │ setosa  │
│ 6   │ 5.4         │ 3.9        │ 1.7         │ 0.4        │ setosa  │
│ 7   │ 4.6         │ 3.4        │ 1.4         │ 0.3        │ setosa  │
│ 8   │ 5.0         │ 3.4        │ 1.5         │ 0.2        │ setosa  │
│ 9   │ 4.4         │ 2.9        │ 1.4         │ 0.2        │ setosa  │
│ 10  │ 4.9         │ 3.1        │ 1.5         │ 0.1        │ setosa  │

Actual data science problems generally do not deal with the artificial randomly generated data or data read through the command line. But they work on data that is loaded from files or any other external source. These files can have data in any format and we may have to process it before loading it to the dataframe.

Julia provides a readtable() function that can be used to read a tabular file in a dataframe.

Although the iris dataset is available in the RDatasets package, we will download the CSV to work with the external datasets. The iris CSV can be downloaded from 

https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/datasets/data/iris.csv.


# Reading the data using dataframe

In [25]:
using DataFrames
using CSV

In [26]:
df_iris_sample = readtable("Iris.csv",separator = ',')

UndefVarError: UndefVarError: readtable not defined

Note: "readtable is deprecated, use CSV.read from the CSV package instead",:readtable

In [27]:
df_iris_sample=DataFrame!(CSV.File("Iris.csv"))

150×6 DataFrame. Omitted printing of 1 columns
│ Row │ Id    │ SepalLengthCm │ SepalWidthCm │ PetalLengthCm │ PetalWidthCm │
│     │ Int64 │ Float64       │ Float64      │ Float64       │ Float64      │
├─────┼───────┼───────────────┼──────────────┼───────────────┼──────────────┤
│ 1   │ 1     │ 5.1           │ 3.5          │ 1.4           │ 0.2          │
│ 2   │ 2     │ 4.9           │ 3.0          │ 1.4           │ 0.2          │
│ 3   │ 3     │ 4.7           │ 3.2          │ 1.3           │ 0.2          │
│ 4   │ 4     │ 4.6           │ 3.1          │ 1.5           │ 0.2          │
│ 5   │ 5     │ 5.0           │ 3.6          │ 1.4           │ 0.2          │
│ 6   │ 6     │ 5.4           │ 3.9          │ 1.7           │ 0.4          │
│ 7   │ 7     │ 4.6           │ 3.4          │ 1.4           │ 0.3          │
│ 8   │ 8     │ 5.0           │ 3.4          │ 1.5           │ 0.2          │
│ 9   │ 9     │ 4.4           │ 2.9          │ 1.4           │ 0.2          │
│ 10  │ 10    │ 4.9           │ 3.1          │ 1.5           │ 0.1          │
⋮
│ 140 │ 140   │ 6.9           │ 3.1          │ 5.4           │ 2.1          │
│ 141 │ 141   │ 6.7           │ 3.1          │ 5.6           │ 2.4          │
│ 142 │ 142   │ 6.9           │ 3.1          │ 5.1           │ 2.3          │
│ 143 │ 143   │ 5.8           │ 2.7          │ 5.1           │ 1.9          │
│ 144 │ 144   │ 6.8           │ 3.2          │ 5.9           │ 2.3          │
│ 145 │ 145   │ 6.7           │ 3.3          │ 5.7           │ 2.5          │
│ 146 │ 146   │ 6.7           │ 3.0          │ 5.2           │ 2.3          │
│ 147 │ 147   │ 6.3           │ 2.5          │ 5.0           │ 1.9          │
│ 148 │ 148   │ 6.5           │ 3.0          │ 5.2           │ 2.0          │
│ 149 │ 149   │ 6.2           │ 3.4          │ 5.4           │ 2.3          │
│ 150 │ 150   │ 5.9           │ 3.0          │ 5.1           │ 1.8          │

In [28]:
#df_iris_sample


It is the same dataset that we used in the previous example, but now we are loading the data from a CSV file.

# Writing the data to a file

We may also want to output our results or transform a dataset and store it in a file. In Julia we do this by using the writetable() function. It is very similar to the readtable() function that we discussed in the last section.

In [29]:
writetable("output_df_iris.csv", df_iris_sample)

UndefVarError: UndefVarError: writetable not defined

Note: writetable() is deprecated from Pkg.update() will give me DataFrames 0.11,use CSV.write() instead.

In [30]:
CSV.write("output_df_iris.csv", df_iris_sample)

"output_df_iris.csv"

# Working with DataFrames

We will follow or inherit some of the traditional strategies to manipulate the data. We will go through these strategies and methods in this section and discuss how and why they are important to data science.


# Understanding DataFrames joins

While working with multiple datasets, we often need to merge the datasets in a particular fashion to make the analysis easier or to use it with a particular function.

We will be using the Road Safety Data published by the Department for Transport, UK, and it is open under the OGL-Open Government Licence.

https://data.gov.uk/dataset/road-accidents-safety-data.

We will be using two datasets:

  1. Road Safety Data - Vehicles 2018
  2. Road Safety Data - Accidents 2018

In [31]:
using DataFrames

In [32]:
DfTRoadSafety_Accidents_2018=DataFrame!(CSV.File("dftRoadSafetyData_Accidents_2018.csv"))

122635×32 DataFrame. Omitted printing of 29 columns
│ Row    │ Accident_Index │ Location_Easting_OSGR │ Location_Northing_OSGR │
│        │ String         │ Union{Missing, Int64} │ Union{Missing, Int64}  │
├────────┼────────────────┼───────────────────────┼────────────────────────┤
│ 1      │ 2018010080971  │ 529150                │ 182270                 │
│ 2      │ 2018010080973  │ 542020                │ 184290                 │
│ 3      │ 2018010080974  │ 531720                │ 182910                 │
│ 4      │ 2018010080981  │ 541450                │ 183220                 │
│ 5      │ 2018010080982  │ 543580                │ 176500                 │
│ 6      │ 2018010080983  │ 526060                │ 194910                 │
│ 7      │ 2018010080986  │ 525050                │ 181050                 │
│ 8      │ 2018010080987  │ 536710                │ 176960                 │
│ 9      │ 2018010080990  │ 517110                │ 186280                 │
│ 10     │ 2018010080992  │ 535450                │ 181190                 │
⋮
│ 122625 │ 2018984114918  │ 311794                │ 580755                 │
│ 122626 │ 2018984115118  │ 336343                │ 584662                 │
│ 122627 │ 2018984115218  │ 303700                │ 611260                 │
│ 122628 │ 2018984115318  │ 308374                │ 599895                 │
│ 122629 │ 2018984115418  │ 308481                │ 601469                 │
│ 122630 │ 2018984115518  │ 308354                │ 599190                 │
│ 122631 │ 2018984115718  │ 307124                │ 594145                 │
│ 122632 │ 2018984115918  │ 313325                │ 581065                 │
│ 122633 │ 2018984116018  │ 319337                │ 574511                 │
│ 122634 │ 2018984116318  │ 318858                │ 566932                 │
│ 122635 │ 2018984116418  │ 316008                │ 568771                 │

In [33]:
first(DfTRoadSafety_Accidents_2018,5)

5×32 DataFrame. Omitted printing of 29 columns
│ Row │ Accident_Index │ Location_Easting_OSGR │ Location_Northing_OSGR │
│     │ String         │ Union{Missing, Int64} │ Union{Missing, Int64}  │
├─────┼────────────────┼───────────────────────┼────────────────────────┤
│ 1   │ 2018010080971  │ 529150                │ 182270                 │
│ 2   │ 2018010080973  │ 542020                │ 184290                 │
│ 3   │ 2018010080974  │ 531720                │ 182910                 │
│ 4   │ 2018010080981  │ 541450                │ 183220                 │
│ 5   │ 2018010080982  │ 543580                │ 176500                 │

In [34]:
names(DfTRoadSafety_Accidents_2018)

32-element Array{String,1}:
 "Accident_Index"
 "Location_Easting_OSGR"
 "Location_Northing_OSGR"
 "Longitude"
 "Latitude"
 "Police_Force"
 "Accident_Severity"
 "Number_of_Vehicles"
 "Number_of_Casualties"
 "Date"
 "Day_of_Week"
 "Time"
 "Local_Authority_(District)"
 ⋮
 "2nd_Road_Class"
 "2nd_Road_Number"
 "Pedestrian_Crossing-Human_Control"
 "Pedestrian_Crossing-Physical_Facilities"
 "Light_Conditions"
 "Weather_Conditions"
 "Road_Surface_Conditions"
 "Special_Conditions_at_Site"
 "Carriageway_Hazards"
 "Urban_or_Rural_Area"
 "Did_Police_Officer_Attend_Scene_of_Accident"
 "LSOA_of_Accident_Location"

In [35]:
DfTRoadSafety_Vehicles_2018=DataFrame!(CSV.File("dftRoadSafetyData_Vehicles_2018.csv"))

226409×23 DataFrame. Omitted printing of 20 columns
│ Row    │ Accident_Index │ Vehicle_Reference │ Vehicle_Type │
│        │ String         │ Int64             │ Int64        │
├────────┼────────────────┼───────────────────┼──────────────┤
│ 1      │ 2018010080971  │ 1                 │ 9            │
│ 2      │ 2018010080971  │ 2                 │ 8            │
│ 3      │ 2018010080973  │ 1                 │ 9            │
│ 4      │ 2018010080974  │ 1                 │ 8            │
│ 5      │ 2018010080974  │ 2                 │ 9            │
│ 6      │ 2018010080981  │ 1                 │ 9            │
│ 7      │ 2018010080981  │ 2                 │ 19           │
│ 8      │ 2018010080982  │ 1                 │ 8            │
│ 9      │ 2018010080982  │ 2                 │ 9            │
│ 10     │ 2018010080983  │ 1                 │ 9            │
⋮
│ 226399 │ 2018984115518  │ 1                 │ 9            │
│ 226400 │ 2018984115518  │ 2                 │ 9            │
│ 226401 │ 2018984115718  │ 1                 │ 9            │
│ 226402 │ 2018984115718  │ 2                 │ 9            │
│ 226403 │ 2018984115918  │ 1                 │ 21           │
│ 226404 │ 2018984115918  │ 2                 │ 9            │
│ 226405 │ 2018984116018  │ 1                 │ 9            │
│ 226406 │ 2018984116018  │ 2                 │ 9            │
│ 226407 │ 2018984116018  │ 3                 │ 9            │
│ 226408 │ 2018984116318  │ 1                 │ 2            │
│ 226409 │ 2018984116418  │ 1                 │ 9            │

The second dataset is loaded into the memory.

In [36]:
full_DfTRoadSafety_2018 =join(DfTRoadSafety_Accidents_2018, DfTRoadSafety_Vehicles_2018, on = :Accident_Index)

┌ Warning: inner joining data frames using join is deprecated, use `innerjoin(df1, df2, on=Accident_Index, makeunique=false, validate=(false, false))` instead
│   caller = ip:0x0
└ @ Core :-1


226409×54 DataFrame. Omitted printing of 51 columns
│ Row    │ Accident_Index │ Location_Easting_OSGR │ Location_Northing_OSGR │
│        │ String         │ Union{Missing, Int64} │ Union{Missing, Int64}  │
├────────┼────────────────┼───────────────────────┼────────────────────────┤
│ 1      │ 2018010080971  │ 529150                │ 182270                 │
│ 2      │ 2018010080971  │ 529150                │ 182270                 │
│ 3      │ 2018010080973  │ 542020                │ 184290                 │
│ 4      │ 2018010080974  │ 531720                │ 182910                 │
│ 5      │ 2018010080974  │ 531720                │ 182910                 │
│ 6      │ 2018010080981  │ 541450                │ 183220                 │
│ 7      │ 2018010080981  │ 541450                │ 183220                 │
│ 8      │ 2018010080982  │ 543580                │ 176500                 │
│ 9      │ 2018010080982  │ 543580                │ 176500                 │
│ 10     │ 2018010080983  │ 526060                │ 194910                 │
⋮
│ 226399 │ 2018984115518  │ 308354                │ 599190                 │
│ 226400 │ 2018984115518  │ 308354                │ 599190                 │
│ 226401 │ 2018984115718  │ 307124                │ 594145                 │
│ 226402 │ 2018984115718  │ 307124                │ 594145                 │
│ 226403 │ 2018984115918  │ 313325                │ 581065                 │
│ 226404 │ 2018984115918  │ 313325                │ 581065                 │
│ 226405 │ 2018984116018  │ 319337                │ 574511                 │
│ 226406 │ 2018984116018  │ 319337                │ 574511                 │
│ 226407 │ 2018984116018  │ 319337                │ 574511                 │
│ 226408 │ 2018984116318  │ 318858                │ 566932                 │
│ 226409 │ 2018984116418  │ 316008                │ 568771                 │

In [37]:
first(full_DfTRoadSafety_2018, 5)

5×54 DataFrame. Omitted printing of 51 columns
│ Row │ Accident_Index │ Location_Easting_OSGR │ Location_Northing_OSGR │
│     │ String         │ Union{Missing, Int64} │ Union{Missing, Int64}  │
├─────┼────────────────┼───────────────────────┼────────────────────────┤
│ 1   │ 2018010080971  │ 529150                │ 182270                 │
│ 2   │ 2018010080971  │ 529150                │ 182270                 │
│ 3   │ 2018010080973  │ 542020                │ 184290                 │
│ 4   │ 2018010080974  │ 531720                │ 182910                 │
│ 5   │ 2018010080974  │ 531720                │ 182910                 │

In [38]:
names(full_DfTRoadSafety_2018)

54-element Array{String,1}:
 "Accident_Index"
 "Location_Easting_OSGR"
 "Location_Northing_OSGR"
 "Longitude"
 "Latitude"
 "Police_Force"
 "Accident_Severity"
 "Number_of_Vehicles"
 "Number_of_Casualties"
 "Date"
 "Day_of_Week"
 "Time"
 "Local_Authority_(District)"
 ⋮
 "1st_Point_of_Impact"
 "Was_Vehicle_Left_Hand_Drive?"
 "Journey_Purpose_of_Driver"
 "Sex_of_Driver"
 "Age_of_Driver"
 "Age_Band_of_Driver"
 "Engine_Capacity_(CC)"
 "Propulsion_Code"
 "Age_of_Vehicle"
 "Driver_IMD_Decile"
 "Driver_Home_Area_Type"
 "Vehicle_IMD_Decile"

The benefit is that the join is really easy to do and is really quick, even over large datasets.
    
We have read about other joins available in relation databases. Julia's DataFrames package provides these joins too:

- Inner join: The output, which is the DataFrame, contains only those rows that have keys in both the dataframes.
- Left join: The output DataFrame has the rows for keys that are present in the first (left) DataFrame, irrespective of them being present in the second (right) DataFrame.
- Right join: The output DataFrame has the rows for keys that are present in the second (right) DataFrame, irrespective of them being present in the first (left) DataFrame.
- Outer join: The output DataFrame has the rows for the keys that are present in the first or second DataFrame, which we are joining.
- Semi join: The output DataFrame has only the rows from the first (left) DataFrame for the keys that are present in both the first (left) and second (right) DataFrames. The output contains only rows from the first DataFrame.
- Anti join: The output DataFrame has the rows for keys that are present in the first (left) DataFrame but rows for the same keys are not present in the second (right) DataFrame. The output contains only rows from the first DataFrame.
- Cross join: The output DataFrame has the rows that are the Cartesian product of the rows from the first DataFrame (left) and the second DataFrame (right).

Let's understand this using a simpler dataset. We will create a dataframe and will apply
different joins on it:

In [39]:
Cities1 = ["Delhi","Amsterdam","Hamburg"][rand(1:3, 10)]
Cities = ["Delhi","Amsterdam","Hamburg"][rand(1:3, 10)]

10-element Array{String,1}:
 "Amsterdam"
 "Delhi"
 "Hamburg"
 "Hamburg"
 "Amsterdam"
 "Hamburg"
 "Delhi"
 "Amsterdam"
 "Amsterdam"
 "Hamburg"

In [40]:
df1 = DataFrame(ID=1:10, City=Cities, RandomValue1=rand(1:10,10))

10×3 DataFrame
│ Row │ ID    │ City      │ RandomValue1 │
│     │ Int64 │ String    │ Int64        │
├─────┼───────┼───────────┼──────────────┤
│ 1   │ 1     │ Amsterdam │ 5            │
│ 2   │ 2     │ Delhi     │ 1            │
│ 3   │ 3     │ Hamburg   │ 2            │
│ 4   │ 4     │ Hamburg   │ 8            │
│ 5   │ 5     │ Amsterdam │ 2            │
│ 6   │ 6     │ Hamburg   │ 1            │
│ 7   │ 7     │ Delhi     │ 10           │
│ 8   │ 8     │ Amsterdam │ 6            │
│ 9   │ 9     │ Amsterdam │ 9            │
│ 10  │ 10    │ Hamburg   │ 8            │

In [41]:
df2 = DataFrame(ID = 1:10, City = Cities, RandomValue2 = rand(100:110, 10))

10×3 DataFrame
│ Row │ ID    │ City      │ RandomValue2 │
│     │ Int64 │ String    │ Int64        │
├─────┼───────┼───────────┼──────────────┤
│ 1   │ 1     │ Amsterdam │ 101          │
│ 2   │ 2     │ Delhi     │ 104          │
│ 3   │ 3     │ Hamburg   │ 107          │
│ 4   │ 4     │ Hamburg   │ 101          │
│ 5   │ 5     │ Amsterdam │ 109          │
│ 6   │ 6     │ Hamburg   │ 109          │
│ 7   │ 7     │ Delhi     │ 106          │
│ 8   │ 8     │ Amsterdam │ 110          │
│ 9   │ 9     │ Amsterdam │ 108          │
│ 10  │ 10    │ Hamburg   │ 103          │

This created two dataframes having 10 rows. The first dataframe, df1, has three columns: ID, City, and RandomValue1. The second dataframe has df2 with three columns: ID, City, and RandomValue2.

This will generate:

In [42]:
full_df1_df2 = join(df1,df2,on = [:ID, :City])

10×4 DataFrame
│ Row │ ID    │ City      │ RandomValue1 │ RandomValue2 │
│     │ Int64 │ String    │ Int64        │ Int64        │
├─────┼───────┼───────────┼──────────────┼──────────────┤
│ 1   │ 1     │ Amsterdam │ 5            │ 101          │
│ 2   │ 2     │ Delhi     │ 1            │ 104          │
│ 3   │ 3     │ Hamburg   │ 2            │ 107          │
│ 4   │ 4     │ Hamburg   │ 8            │ 101          │
│ 5   │ 5     │ Amsterdam │ 2            │ 109          │
│ 6   │ 6     │ Hamburg   │ 1            │ 109          │
│ 7   │ 7     │ Delhi     │ 10           │ 106          │
│ 8   │ 8     │ Amsterdam │ 6            │ 110          │
│ 9   │ 9     │ Amsterdam │ 9            │ 108          │
│ 10  │ 10    │ Hamburg   │ 8            │ 103          │

Other joins can also be applied using the kind argument. Let's go through our old dataset of accidents and vehicles.

In [43]:
right_DfTRoadSafety_2018 = join(DfTRoadSafety_Accidents_2018,DfTRoadSafety_Vehicles_2018, on = :Accident_Index, kind = :right)

226409×54 DataFrame. Omitted printing of 51 columns
│ Row    │ Accident_Index │ Location_Easting_OSGR │ Location_Northing_OSGR │
│        │ String?        │ Union{Missing, Int64} │ Union{Missing, Int64}  │
├────────┼────────────────┼───────────────────────┼────────────────────────┤
│ 1      │ 2018010080971  │ 529150                │ 182270                 │
│ 2      │ 2018010080971  │ 529150                │ 182270                 │
│ 3      │ 2018010080973  │ 542020                │ 184290                 │
│ 4      │ 2018010080974  │ 531720                │ 182910                 │
│ 5      │ 2018010080974  │ 531720                │ 182910                 │
│ 6      │ 2018010080981  │ 541450                │ 183220                 │
│ 7      │ 2018010080981  │ 541450                │ 183220                 │
│ 8      │ 2018010080982  │ 543580                │ 176500                 │
│ 9      │ 2018010080982  │ 543580                │ 176500                 │
│ 10     │ 2018010080983  │ 526060                │ 194910                 │
⋮
│ 226399 │ 2018984115518  │ 308354                │ 599190                 │
│ 226400 │ 2018984115518  │ 308354                │ 599190                 │
│ 226401 │ 2018984115718  │ 307124                │ 594145                 │
│ 226402 │ 2018984115718  │ 307124                │ 594145                 │
│ 226403 │ 2018984115918  │ 313325                │ 581065                 │
│ 226404 │ 2018984115918  │ 313325                │ 581065                 │
│ 226405 │ 2018984116018  │ 319337                │ 574511                 │
│ 226406 │ 2018984116018  │ 319337                │ 574511                 │
│ 226407 │ 2018984116018  │ 319337                │ 574511                 │
│ 226408 │ 2018984116318  │ 318858                │ 566932                 │
│ 226409 │ 2018984116418  │ 316008                │ 568771                 │

In [44]:
inner_DfTRoadSafety_2018 = join(DfTRoadSafety_Accidents_2018, DfTRoadSafety_Vehicles_2018, on = :Accident_Index, kind = :inner)

226409×54 DataFrame. Omitted printing of 51 columns
│ Row    │ Accident_Index │ Location_Easting_OSGR │ Location_Northing_OSGR │
│        │ String         │ Union{Missing, Int64} │ Union{Missing, Int64}  │
├────────┼────────────────┼───────────────────────┼────────────────────────┤
│ 1      │ 2018010080971  │ 529150                │ 182270                 │
│ 2      │ 2018010080971  │ 529150                │ 182270                 │
│ 3      │ 2018010080973  │ 542020                │ 184290                 │
│ 4      │ 2018010080974  │ 531720                │ 182910                 │
│ 5      │ 2018010080974  │ 531720                │ 182910                 │
│ 6      │ 2018010080981  │ 541450                │ 183220                 │
│ 7      │ 2018010080981  │ 541450                │ 183220                 │
│ 8      │ 2018010080982  │ 543580                │ 176500                 │
│ 9      │ 2018010080982  │ 543580                │ 176500                 │
│ 10     │ 2018010080983  │ 526060                │ 194910                 │
⋮
│ 226399 │ 2018984115518  │ 308354                │ 599190                 │
│ 226400 │ 2018984115518  │ 308354                │ 599190                 │
│ 226401 │ 2018984115718  │ 307124                │ 594145                 │
│ 226402 │ 2018984115718  │ 307124                │ 594145                 │
│ 226403 │ 2018984115918  │ 313325                │ 581065                 │
│ 226404 │ 2018984115918  │ 313325                │ 581065                 │
│ 226405 │ 2018984116018  │ 319337                │ 574511                 │
│ 226406 │ 2018984116018  │ 319337                │ 574511                 │
│ 226407 │ 2018984116018  │ 319337                │ 574511                 │
│ 226408 │ 2018984116318  │ 318858                │ 566932                 │
│ 226409 │ 2018984116418  │ 316008                │ 568771                 │

In [45]:
outer_DfTRoadSafety_2018 = join(DfTRoadSafety_Accidents_2018, DfTRoadSafety_Vehicles_2018, on = :Accident_Index, kind = :outer)

226409×54 DataFrame. Omitted printing of 51 columns
│ Row    │ Accident_Index │ Location_Easting_OSGR │ Location_Northing_OSGR │
│        │ String?        │ Union{Missing, Int64} │ Union{Missing, Int64}  │
├────────┼────────────────┼───────────────────────┼────────────────────────┤
│ 1      │ 2018010080971  │ 529150                │ 182270                 │
│ 2      │ 2018010080971  │ 529150                │ 182270                 │
│ 3      │ 2018010080973  │ 542020                │ 184290                 │
│ 4      │ 2018010080974  │ 531720                │ 182910                 │
│ 5      │ 2018010080974  │ 531720                │ 182910                 │
│ 6      │ 2018010080981  │ 541450                │ 183220                 │
│ 7      │ 2018010080981  │ 541450                │ 183220                 │
│ 8      │ 2018010080982  │ 543580                │ 176500                 │
│ 9      │ 2018010080982  │ 543580                │ 176500                 │
│ 10     │ 2018010080983  │ 526060                │ 194910                 │
⋮
│ 226399 │ 2018984115518  │ 308354                │ 599190                 │
│ 226400 │ 2018984115518  │ 308354                │ 599190                 │
│ 226401 │ 2018984115718  │ 307124                │ 594145                 │
│ 226402 │ 2018984115718  │ 307124                │ 594145                 │
│ 226403 │ 2018984115918  │ 313325                │ 581065                 │
│ 226404 │ 2018984115918  │ 313325                │ 581065                 │
│ 226405 │ 2018984116018  │ 319337                │ 574511                 │
│ 226406 │ 2018984116018  │ 319337                │ 574511                 │
│ 226407 │ 2018984116018  │ 319337                │ 574511                 │
│ 226408 │ 2018984116318  │ 318858                │ 566932                 │
│ 226409 │ 2018984116418  │ 316008                │ 568771                 │

In [46]:
semi_DfTRoadSafety_2018 = join(DfTRoadSafety_Accidents_2018, DfTRoadSafety_Vehicles_2018, on = :Accident_Index, kind = :semi)

122635×32 DataFrame. Omitted printing of 29 columns
│ Row    │ Accident_Index │ Location_Easting_OSGR │ Location_Northing_OSGR │
│        │ String         │ Union{Missing, Int64} │ Union{Missing, Int64}  │
├────────┼────────────────┼───────────────────────┼────────────────────────┤
│ 1      │ 2018010080971  │ 529150                │ 182270                 │
│ 2      │ 2018010080973  │ 542020                │ 184290                 │
│ 3      │ 2018010080974  │ 531720                │ 182910                 │
│ 4      │ 2018010080981  │ 541450                │ 183220                 │
│ 5      │ 2018010080982  │ 543580                │ 176500                 │
│ 6      │ 2018010080983  │ 526060                │ 194910                 │
│ 7      │ 2018010080986  │ 525050                │ 181050                 │
│ 8      │ 2018010080987  │ 536710                │ 176960                 │
│ 9      │ 2018010080990  │ 517110                │ 186280                 │
│ 10     │ 2018010080992  │ 535450                │ 181190                 │
⋮
│ 122625 │ 2018984114918  │ 311794                │ 580755                 │
│ 122626 │ 2018984115118  │ 336343                │ 584662                 │
│ 122627 │ 2018984115218  │ 303700                │ 611260                 │
│ 122628 │ 2018984115318  │ 308374                │ 599895                 │
│ 122629 │ 2018984115418  │ 308481                │ 601469                 │
│ 122630 │ 2018984115518  │ 308354                │ 599190                 │
│ 122631 │ 2018984115718  │ 307124                │ 594145                 │
│ 122632 │ 2018984115918  │ 313325                │ 581065                 │
│ 122633 │ 2018984116018  │ 319337                │ 574511                 │
│ 122634 │ 2018984116318  │ 318858                │ 566932                 │
│ 122635 │ 2018984116418  │ 316008                │ 568771                 │

In [47]:
anti_DfTRoadSafety_2018 = join(DfTRoadSafety_Accidents_2018, DfTRoadSafety_Vehicles_2018, on = :Accident_Index, kind = :anti)

0×32 DataFrame

# The Split-Apply-Combine strategy

This is needed when a dataset contains a large number of columns and for some operations all the columns are not necessary. It is better to split the dataset and then apply the necessary functions; and we can always put the dataset back together.

This is done using the by function by takes three arguments:
- DataFrame (this is the dataframe that we would be splitting)
- The column name (or numerical index) on which the DataFrame would be split 
- A function that can be applied on every subset of the DataFrame


In [48]:
by(DfTRoadSafety_Accidents_2018, :Location_Northing_OSGR, size)

┌ Warning: `by(d::AbstractDataFrame, cols::Any, f::Base.Callable; sort::Bool = false, skipmissing::Bool = false)` is deprecated, use `combine(f, groupby(d, cols, sort = sort, skipmissing = skipmissing))` instead.
│   caller = top-level scope at In[48]:1
└ @ Core In[48]:1


86864×2 DataFrame
│ Row   │ Location_Northing_OSGR │ x1       │
│       │ Union{Missing, Int64}  │ Tuple…   │
├───────┼────────────────────────┼──────────┤
│ 1     │ 182270                 │ (31, 32) │
│ 2     │ 184290                 │ (13, 32) │
│ 3     │ 182910                 │ (19, 32) │
│ 4     │ 183220                 │ (10, 32) │
│ 5     │ 176500                 │ (18, 32) │
│ 6     │ 194910                 │ (2, 32)  │
│ 7     │ 181050                 │ (23, 32) │
│ 8     │ 176960                 │ (26, 32) │
│ 9     │ 186280                 │ (12, 32) │
│ 10    │ 181190                 │ (21, 32) │
⋮
│ 86854 │ 567300                 │ (1, 32)  │
│ 86855 │ 580755                 │ (1, 32)  │
│ 86856 │ 584662                 │ (1, 32)  │
│ 86857 │ 611260                 │ (1, 32)  │
│ 86858 │ 599895                 │ (1, 32)  │
│ 86859 │ 601469                 │ (1, 32)  │
│ 86860 │ 599190                 │ (1, 32)  │
│ 86861 │ 594145                 │ (1, 32)  │
│ 86862 │ 581065                 │ (1, 32)  │
│ 86863 │ 574511                 │ (1, 32)  │
│ 86864 │ 566932                 │ (1, 32)  │

The aggregate() function provides an alternative to apply the Split-Apply-Combine strategy. The aggregate() function uses the same three arguments:
- DataFrame (this is the DataFrame that we would be splitting)
- The column name (or numerical index) on which the DataFrame would be split 
- A function that can be applied on the every subset of the DataFrame

# Reshaping the data

The use case may require data to be in a different shape than we currently have. To facilitate this, Julia provides reshaping of the data.


Let's use the same dataset that we were using, but before that let's check the size of the dataset:

In [49]:
size(DfTRoadSafety_Accidents_2018)

(122635, 32)

We will import RDatasets and DataFrames 

In [50]:
using RDatasets, DataFrames

Then, we will load the iris dataset into a DataFrame. We can see that the dataset has 150 rows and 5 columns:

In [51]:
iris_dataframe = dataset("datasets", "iris")

150×5 DataFrame
│ Row │ SepalLength │ SepalWidth │ PetalLength │ PetalWidth │ Species   │
│     │ Float64     │ Float64    │ Float64     │ Float64    │ Cat…      │
├─────┼─────────────┼────────────┼─────────────┼────────────┼───────────┤
│ 1   │ 5.1         │ 3.5        │ 1.4         │ 0.2        │ setosa    │
│ 2   │ 4.9         │ 3.0        │ 1.4         │ 0.2        │ setosa    │
│ 3   │ 4.7         │ 3.2        │ 1.3         │ 0.2        │ setosa    │
│ 4   │ 4.6         │ 3.1        │ 1.5         │ 0.2        │ setosa    │
│ 5   │ 5.0         │ 3.6        │ 1.4         │ 0.2        │ setosa    │
│ 6   │ 5.4         │ 3.9        │ 1.7         │ 0.4        │ setosa    │
│ 7   │ 4.6         │ 3.4        │ 1.4         │ 0.3        │ setosa    │
│ 8   │ 5.0         │ 3.4        │ 1.5         │ 0.2        │ setosa    │
│ 9   │ 4.4         │ 2.9        │ 1.4         │ 0.2        │ setosa    │
│ 10  │ 4.9         │ 3.1        │ 1.5         │ 0.1        │ setosa    │
⋮
│ 140 │ 6.9         │ 3.1        │ 5.4         │ 2.1        │ virginica │
│ 141 │ 6.7         │ 3.1        │ 5.6         │ 2.4        │ virginica │
│ 142 │ 6.9         │ 3.1        │ 5.1         │ 2.3        │ virginica │
│ 143 │ 5.8         │ 2.7        │ 5.1         │ 1.9        │ virginica │
│ 144 │ 6.8         │ 3.2        │ 5.9         │ 2.3        │ virginica │
│ 145 │ 6.7         │ 3.3        │ 5.7         │ 2.5        │ virginica │
│ 146 │ 6.7         │ 3.0        │ 5.2         │ 2.3        │ virginica │
│ 147 │ 6.3         │ 2.5        │ 5.0         │ 1.9        │ virginica │
│ 148 │ 6.5         │ 3.0        │ 5.2         │ 2.0        │ virginica │
│ 149 │ 6.2         │ 3.4        │ 5.4         │ 2.3        │ virginica │
│ 150 │ 5.9         │ 3.0        │ 5.1         │ 1.8        │ virginica │

Now we use the stack() function to reshape the dataset. Let's use it without any arguments except the DataFrame.

In [52]:
iris_stackdf = stack(iris_dataframe)

600×3 DataFrame
│ Row │ Species   │ variable    │ value   │
│     │ Cat…      │ Cat…        │ Float64 │
├─────┼───────────┼─────────────┼─────────┤
│ 1   │ setosa    │ SepalLength │ 5.1     │
│ 2   │ setosa    │ SepalLength │ 4.9     │
│ 3   │ setosa    │ SepalLength │ 4.7     │
│ 4   │ setosa    │ SepalLength │ 4.6     │
│ 5   │ setosa    │ SepalLength │ 5.0     │
│ 6   │ setosa    │ SepalLength │ 5.4     │
│ 7   │ setosa    │ SepalLength │ 4.6     │
│ 8   │ setosa    │ SepalLength │ 5.0     │
│ 9   │ setosa    │ SepalLength │ 4.4     │
│ 10  │ setosa    │ SepalLength │ 4.9     │
⋮
│ 590 │ virginica │ PetalWidth  │ 2.1     │
│ 591 │ virginica │ PetalWidth  │ 2.4     │
│ 592 │ virginica │ PetalWidth  │ 2.3     │
│ 593 │ virginica │ PetalWidth  │ 1.9     │
│ 594 │ virginica │ PetalWidth  │ 2.3     │
│ 595 │ virginica │ PetalWidth  │ 2.5     │
│ 596 │ virginica │ PetalWidth  │ 2.3     │
│ 597 │ virginica │ PetalWidth  │ 1.9     │
│ 598 │ virginica │ PetalWidth  │ 2.0     │
│ 599 │ virginica │ PetalWidth  │ 2.3     │
│ 600 │ virginica │ PetalWidth  │ 1.8     │

We can see that our dataset has been stacked. Here we have stacked all the columns. We can also provide specific columns to stack:

In [53]:
iris_dataframe[:id] = 1:size(iris_dataframe, 1)

┌ Warning: `setindex!(df::DataFrame, v::AbstractVector, col_ind::ColumnIndex)` is deprecated, use `begin
│     df[!, col_ind] = v
│     df
│ end` instead.
│   caller = top-level scope at In[53]:1
└ @ Core In[53]:1


1:150

In [54]:
iris_stack = stack(iris_dataframe)


600×4 DataFrame
│ Row │ Species   │ id    │ variable    │ value   │
│     │ Cat…      │ Int64 │ Cat…        │ Float64 │
├─────┼───────────┼───────┼─────────────┼─────────┤
│ 1   │ setosa    │ 1     │ SepalLength │ 5.1     │
│ 2   │ setosa    │ 2     │ SepalLength │ 4.9     │
│ 3   │ setosa    │ 3     │ SepalLength │ 4.7     │
│ 4   │ setosa    │ 4     │ SepalLength │ 4.6     │
│ 5   │ setosa    │ 5     │ SepalLength │ 5.0     │
│ 6   │ setosa    │ 6     │ SepalLength │ 5.4     │
│ 7   │ setosa    │ 7     │ SepalLength │ 4.6     │
│ 8   │ setosa    │ 8     │ SepalLength │ 5.0     │
│ 9   │ setosa    │ 9     │ SepalLength │ 4.4     │
│ 10  │ setosa    │ 10    │ SepalLength │ 4.9     │
⋮
│ 590 │ virginica │ 140   │ PetalWidth  │ 2.1     │
│ 591 │ virginica │ 141   │ PetalWidth  │ 2.4     │
│ 592 │ virginica │ 142   │ PetalWidth  │ 2.3     │
│ 593 │ virginica │ 143   │ PetalWidth  │ 1.9     │
│ 594 │ virginica │ 144   │ PetalWidth  │ 2.3     │
│ 595 │ virginica │ 145   │ PetalWidth  │ 2.5     │
│ 596 │ virginica │ 146   │ PetalWidth  │ 2.3     │
│ 597 │ virginica │ 147   │ PetalWidth  │ 1.9     │
│ 598 │ virginica │ 148   │ PetalWidth  │ 2.0     │
│ 599 │ virginica │ 149   │ PetalWidth  │ 2.3     │
│ 600 │ virginica │ 150   │ PetalWidth  │ 1.8     │

In [55]:
size(iris_stack)

(600, 4)

In [56]:
head(iris_stack)

┌ Warning: `head(df::AbstractDataFrame)` is deprecated, use `first(df, 6)` instead.
│   caller = top-level scope at In[56]:1
└ @ Core In[56]:1


6×4 DataFrame
│ Row │ Species │ id    │ variable    │ value   │
│     │ Cat…    │ Int64 │ Cat…        │ Float64 │
├─────┼─────────┼───────┼─────────────┼─────────┤
│ 1   │ setosa  │ 1     │ SepalLength │ 5.1     │
│ 2   │ setosa  │ 2     │ SepalLength │ 4.9     │
│ 3   │ setosa  │ 3     │ SepalLength │ 4.7     │
│ 4   │ setosa  │ 4     │ SepalLength │ 4.6     │
│ 5   │ setosa  │ 5     │ SepalLength │ 5.0     │
│ 6   │ setosa  │ 6     │ SepalLength │ 5.4     │

We can see that there is a new column :id. That's the identifier of the stacked dataframe. Its value is repeated the number of times the rows are repeated.

# Sorting a dataset

Sorting is one of the most used techniques in data analysis. Sorting is facilitated in Julia by calling the sort or sort! function.
    
The difference between the sort and sort! is that sort! works in-place, which sorts the actual array rather than creating a copy.

Let's use the sort! function on the iris dataset:

In [57]:
iris_dataframe1 = dataset("datasets", "iris")

150×5 DataFrame
│ Row │ SepalLength │ SepalWidth │ PetalLength │ PetalWidth │ Species   │
│     │ Float64     │ Float64    │ Float64     │ Float64    │ Cat…      │
├─────┼─────────────┼────────────┼─────────────┼────────────┼───────────┤
│ 1   │ 5.1         │ 3.5        │ 1.4         │ 0.2        │ setosa    │
│ 2   │ 4.9         │ 3.0        │ 1.4         │ 0.2        │ setosa    │
│ 3   │ 4.7         │ 3.2        │ 1.3         │ 0.2        │ setosa    │
│ 4   │ 4.6         │ 3.1        │ 1.5         │ 0.2        │ setosa    │
│ 5   │ 5.0         │ 3.6        │ 1.4         │ 0.2        │ setosa    │
│ 6   │ 5.4         │ 3.9        │ 1.7         │ 0.4        │ setosa    │
│ 7   │ 4.6         │ 3.4        │ 1.4         │ 0.3        │ setosa    │
│ 8   │ 5.0         │ 3.4        │ 1.5         │ 0.2        │ setosa    │
│ 9   │ 4.4         │ 2.9        │ 1.4         │ 0.2        │ setosa    │
│ 10  │ 4.9         │ 3.1        │ 1.5         │ 0.1        │ setosa    │
⋮
│ 140 │ 6.9         │ 3.1        │ 5.4         │ 2.1        │ virginica │
│ 141 │ 6.7         │ 3.1        │ 5.6         │ 2.4        │ virginica │
│ 142 │ 6.9         │ 3.1        │ 5.1         │ 2.3        │ virginica │
│ 143 │ 5.8         │ 2.7        │ 5.1         │ 1.9        │ virginica │
│ 144 │ 6.8         │ 3.2        │ 5.9         │ 2.3        │ virginica │
│ 145 │ 6.7         │ 3.3        │ 5.7         │ 2.5        │ virginica │
│ 146 │ 6.7         │ 3.0        │ 5.2         │ 2.3        │ virginica │
│ 147 │ 6.3         │ 2.5        │ 5.0         │ 1.9        │ virginica │
│ 148 │ 6.5         │ 3.0        │ 5.2         │ 2.0        │ virginica │
│ 149 │ 6.2         │ 3.4        │ 5.4         │ 2.3        │ virginica │
│ 150 │ 5.9         │ 3.0        │ 5.1         │ 1.8        │ virginica │

In [58]:
sort!(iris_dataframe1)

150×5 DataFrame
│ Row │ SepalLength │ SepalWidth │ PetalLength │ PetalWidth │ Species   │
│     │ Float64     │ Float64    │ Float64     │ Float64    │ Cat…      │
├─────┼─────────────┼────────────┼─────────────┼────────────┼───────────┤
│ 1   │ 4.3         │ 3.0        │ 1.1         │ 0.1        │ setosa    │
│ 2   │ 4.4         │ 2.9        │ 1.4         │ 0.2        │ setosa    │
│ 3   │ 4.4         │ 3.0        │ 1.3         │ 0.2        │ setosa    │
│ 4   │ 4.4         │ 3.2        │ 1.3         │ 0.2        │ setosa    │
│ 5   │ 4.5         │ 2.3        │ 1.3         │ 0.3        │ setosa    │
│ 6   │ 4.6         │ 3.1        │ 1.5         │ 0.2        │ setosa    │
│ 7   │ 4.6         │ 3.2        │ 1.4         │ 0.2        │ setosa    │
│ 8   │ 4.6         │ 3.4        │ 1.4         │ 0.3        │ setosa    │
│ 9   │ 4.6         │ 3.6        │ 1.0         │ 0.2        │ setosa    │
│ 10  │ 4.7         │ 3.2        │ 1.3         │ 0.2        │ setosa    │
⋮
│ 140 │ 7.2         │ 3.0        │ 5.8         │ 1.6        │ virginica │
│ 141 │ 7.2         │ 3.2        │ 6.0         │ 1.8        │ virginica │
│ 142 │ 7.2         │ 3.6        │ 6.1         │ 2.5        │ virginica │
│ 143 │ 7.3         │ 2.9        │ 6.3         │ 1.8        │ virginica │
│ 144 │ 7.4         │ 2.8        │ 6.1         │ 1.9        │ virginica │
│ 145 │ 7.6         │ 3.0        │ 6.6         │ 2.1        │ virginica │
│ 146 │ 7.7         │ 2.6        │ 6.9         │ 2.3        │ virginica │
│ 147 │ 7.7         │ 2.8        │ 6.7         │ 2.0        │ virginica │
│ 148 │ 7.7         │ 3.0        │ 6.1         │ 2.3        │ virginica │
│ 149 │ 7.7         │ 3.8        │ 6.7         │ 2.2        │ virginica │
│ 150 │ 7.9         │ 3.8        │ 6.4         │ 2.0        │ virginica │

We can see that the columns are not sorted according to [:SepalLength, :SepalWidth, :PetalLength, :PetalWidth]. But these are actually sorted according to the :Species column.

The sorting function takes some arguments and provides a few features. For example, to sort in reverse, we have:


In [59]:
sort!(iris_dataframe1, rev = true)


150×5 DataFrame
│ Row │ SepalLength │ SepalWidth │ PetalLength │ PetalWidth │ Species   │
│     │ Float64     │ Float64    │ Float64     │ Float64    │ Cat…      │
├─────┼─────────────┼────────────┼─────────────┼────────────┼───────────┤
│ 1   │ 7.9         │ 3.8        │ 6.4         │ 2.0        │ virginica │
│ 2   │ 7.7         │ 3.8        │ 6.7         │ 2.2        │ virginica │
│ 3   │ 7.7         │ 3.0        │ 6.1         │ 2.3        │ virginica │
│ 4   │ 7.7         │ 2.8        │ 6.7         │ 2.0        │ virginica │
│ 5   │ 7.7         │ 2.6        │ 6.9         │ 2.3        │ virginica │
│ 6   │ 7.6         │ 3.0        │ 6.6         │ 2.1        │ virginica │
│ 7   │ 7.4         │ 2.8        │ 6.1         │ 1.9        │ virginica │
│ 8   │ 7.3         │ 2.9        │ 6.3         │ 1.8        │ virginica │
│ 9   │ 7.2         │ 3.6        │ 6.1         │ 2.5        │ virginica │
│ 10  │ 7.2         │ 3.2        │ 6.0         │ 1.8        │ virginica │
⋮
│ 140 │ 4.7         │ 3.2        │ 1.6         │ 0.2        │ setosa    │
│ 141 │ 4.7         │ 3.2        │ 1.3         │ 0.2        │ setosa    │
│ 142 │ 4.6         │ 3.6        │ 1.0         │ 0.2        │ setosa    │
│ 143 │ 4.6         │ 3.4        │ 1.4         │ 0.3        │ setosa    │
│ 144 │ 4.6         │ 3.2        │ 1.4         │ 0.2        │ setosa    │
│ 145 │ 4.6         │ 3.1        │ 1.5         │ 0.2        │ setosa    │
│ 146 │ 4.5         │ 2.3        │ 1.3         │ 0.3        │ setosa    │
│ 147 │ 4.4         │ 3.2        │ 1.3         │ 0.2        │ setosa    │
│ 148 │ 4.4         │ 3.0        │ 1.3         │ 0.2        │ setosa    │
│ 149 │ 4.4         │ 2.9        │ 1.4         │ 0.2        │ setosa    │
│ 150 │ 4.3         │ 3.0        │ 1.1         │ 0.1        │ setosa    │

To sort some specific columns, we have:

In [60]:
sort!(iris_dataframe1, [:SepalLength, :PetalLength])


150×5 DataFrame
│ Row │ SepalLength │ SepalWidth │ PetalLength │ PetalWidth │ Species   │
│     │ Float64     │ Float64    │ Float64     │ Float64    │ Cat…      │
├─────┼─────────────┼────────────┼─────────────┼────────────┼───────────┤
│ 1   │ 4.3         │ 3.0        │ 1.1         │ 0.1        │ setosa    │
│ 2   │ 4.4         │ 3.2        │ 1.3         │ 0.2        │ setosa    │
│ 3   │ 4.4         │ 3.0        │ 1.3         │ 0.2        │ setosa    │
│ 4   │ 4.4         │ 2.9        │ 1.4         │ 0.2        │ setosa    │
│ 5   │ 4.5         │ 2.3        │ 1.3         │ 0.3        │ setosa    │
│ 6   │ 4.6         │ 3.6        │ 1.0         │ 0.2        │ setosa    │
│ 7   │ 4.6         │ 3.4        │ 1.4         │ 0.3        │ setosa    │
│ 8   │ 4.6         │ 3.2        │ 1.4         │ 0.2        │ setosa    │
│ 9   │ 4.6         │ 3.1        │ 1.5         │ 0.2        │ setosa    │
│ 10  │ 4.7         │ 3.2        │ 1.3         │ 0.2        │ setosa    │
⋮
│ 140 │ 7.2         │ 3.0        │ 5.8         │ 1.6        │ virginica │
│ 141 │ 7.2         │ 3.2        │ 6.0         │ 1.8        │ virginica │
│ 142 │ 7.2         │ 3.6        │ 6.1         │ 2.5        │ virginica │
│ 143 │ 7.3         │ 2.9        │ 6.3         │ 1.8        │ virginica │
│ 144 │ 7.4         │ 2.8        │ 6.1         │ 1.9        │ virginica │
│ 145 │ 7.6         │ 3.0        │ 6.6         │ 2.1        │ virginica │
│ 146 │ 7.7         │ 3.0        │ 6.1         │ 2.3        │ virginica │
│ 147 │ 7.7         │ 3.8        │ 6.7         │ 2.2        │ virginica │
│ 148 │ 7.7         │ 2.8        │ 6.7         │ 2.0        │ virginica │
│ 149 │ 7.7         │ 2.6        │ 6.9         │ 2.3        │ virginica │
│ 150 │ 7.9         │ 3.8        │ 6.4         │ 2.0        │ virginica │

We can also use the by function with sort! to apply another function on the DataFrame or the single column.

In [61]:
sort!(iris_dataframe1, [order(:SepalLength), order(:PetalLength, rev=true)])

150×5 DataFrame
│ Row │ SepalLength │ SepalWidth │ PetalLength │ PetalWidth │ Species   │
│     │ Float64     │ Float64    │ Float64     │ Float64    │ Cat…      │
├─────┼─────────────┼────────────┼─────────────┼────────────┼───────────┤
│ 1   │ 4.3         │ 3.0        │ 1.1         │ 0.1        │ setosa    │
│ 2   │ 4.4         │ 2.9        │ 1.4         │ 0.2        │ setosa    │
│ 3   │ 4.4         │ 3.2        │ 1.3         │ 0.2        │ setosa    │
│ 4   │ 4.4         │ 3.0        │ 1.3         │ 0.2        │ setosa    │
│ 5   │ 4.5         │ 2.3        │ 1.3         │ 0.3        │ setosa    │
│ 6   │ 4.6         │ 3.1        │ 1.5         │ 0.2        │ setosa    │
│ 7   │ 4.6         │ 3.4        │ 1.4         │ 0.3        │ setosa    │
│ 8   │ 4.6         │ 3.2        │ 1.4         │ 0.2        │ setosa    │
│ 9   │ 4.6         │ 3.6        │ 1.0         │ 0.2        │ setosa    │
│ 10  │ 4.7         │ 3.2        │ 1.6         │ 0.2        │ setosa    │
⋮
│ 140 │ 7.2         │ 3.6        │ 6.1         │ 2.5        │ virginica │
│ 141 │ 7.2         │ 3.2        │ 6.0         │ 1.8        │ virginica │
│ 142 │ 7.2         │ 3.0        │ 5.8         │ 1.6        │ virginica │
│ 143 │ 7.3         │ 2.9        │ 6.3         │ 1.8        │ virginica │
│ 144 │ 7.4         │ 2.8        │ 6.1         │ 1.9        │ virginica │
│ 145 │ 7.6         │ 3.0        │ 6.6         │ 2.1        │ virginica │
│ 146 │ 7.7         │ 2.6        │ 6.9         │ 2.3        │ virginica │
│ 147 │ 7.7         │ 3.8        │ 6.7         │ 2.2        │ virginica │
│ 148 │ 7.7         │ 2.8        │ 6.7         │ 2.0        │ virginica │
│ 149 │ 7.7         │ 3.0        │ 6.1         │ 2.3        │ virginica │
│ 150 │ 7.9         │ 3.8        │ 6.4         │ 2.0        │ virginica │

Order is used to specify ordering a specific column amongst a set of columns.

# Formula – a special data type for mathematical expressions

Data science involves various statistical formulas to get insights from data. The creation and application of these formulas is one of the core processes of data science. It maps input variables with some function and mathematical expression to an output.


Julia facilitates this by providing a formula type in the DataFrame package, which is used with the symbol ~. ~ is a binary operator.

In [94]:
formulaX = @formula(A ~ B + C)

FormulaTerm
Response:
  A(unknown)
Predictors:
  B(unknown)
  C(unknown)

Create a dataframe with random values:

In [95]:
df = DataFrame(A = randn(5), B = randn(5), C = randn(5))


5×3 DataFrame
│ Row │ A           │ B         │ C         │
│     │ Float64     │ Float64   │ Float64   │
├─────┼─────────────┼───────────┼───────────┤
│ 1   │ -0.0133479  │ -0.669064 │ -0.135336 │
│ 2   │ -2.25753    │ 0.671792  │ -0.915382 │
│ 3   │ -0.00891443 │ 0.68215   │ 0.311072  │
│ 4   │ -0.356262   │ 0.0787835 │ 0.743092  │
│ 5   │ 0.664042    │ -0.664597 │ -0.319786 │

Use formula to transform it into a ModelFrame object:

In [98]:
Random_modelFrame=ModelFrame(@formula(A ~ B + C), df)

ModelFrame{NamedTuple{(:A, :B, :C),Tuple{Array{Float64,1},Array{Float64,1},Array{Float64,1}}},StatisticalModel}(A ~ 1 + B + C, StatsModels.Schema with 3 entries:
  A => A
  C => C
  B => B
, (A = [-0.013347940250632538, -2.257527706873076, -0.008914425629109145, -0.356261673270087, 0.6640420407642994], B = [-0.6690636983849988, 0.6717920390225154, 0.6821503297201112, 0.0787834656552427, -0.6645971826739892], C = [-0.1353358215380402, -0.9153820866448469, 0.3110720327385292, 0.7430924302027259, -0.31978580285282965]), StatisticalModel)

Creating a ModelMatrix from a ModelFrame is quite easy:

In [99]:
mm = ModelMatrix(ModelFrame(@formula(A ~ B + C), df))

ModelMatrix{Array{Float64,2}}([1.0 -0.6690636983849988 -0.1353358215380402; 1.0 0.6717920390225154 -0.9153820866448469; … ; 1.0 0.0787834656552427 0.7430924302027259; 1.0 -0.6645971826739892 -0.31978580285282965], [1, 2, 3])

In [100]:
mm = ModelMatrix(ModelFrame(@formula(C ~ A*B), df))

ModelMatrix{Array{Float64,2}}([1.0 -0.013347940250632538 -0.6690636983849988 0.008930622269910194; 1.0 -2.257527706873076 0.6717920390225154 -1.5165891413500872; … ; 1.0 -0.356261673270087 0.0787834656552427 -0.028067529300353198; 1.0 0.6640420407642994 -0.6645971826739892 -0.44132046946903963], [1, 2, 3, 4])

In [101]:
mm = ModelMatrix(ModelFrame(@formula(C ~ A*B), df, contrasts = Dict(:A => HelmertCoding())))

ModelMatrix{Array{Float64,2}}([1.0 0.0 … 0.6690636983849988 0.6690636983849988; 1.0 -1.0 … -0.6717920390225154 -0.6717920390225154; … ; 1.0 1.0 … -0.0787834656552427 -0.0787834656552427; 1.0 0.0 … -0.0 -2.6583887306959566], [1, 2, 2, 2, 2, 3, 4, 4, 4, 4])

There is an extra column containing only value = 1.0. It is used in a regression model to fit an intercept term.

http://juliadata.github.io/DataFrames.jl/v0.9/man/formulas/

# Pooling data

To analyze huge datasets efficiently, PooledDataArray is used. DataArray uses an encoding that represents a full string for every entry of a vector. This is not very efficient, especially for large datasets and memory-intensive algorithms.

Our use case more often deals with factors involving a small number of levels:

In [105]:
df = DataFrame(A = [1, 1, 1, 2, 2, 2],
               B = ["X", "X", "X", "Y", "Y", "Y"])


6×2 DataFrame
│ Row │ A     │ B      │
│     │ Int64 │ String │
├─────┼───────┼────────┤
│ 1   │ 1     │ X      │
│ 2   │ 1     │ X      │
│ 3   │ 1     │ X      │
│ 4   │ 2     │ Y      │
│ 5   │ 2     │ Y      │
│ 6   │ 2     │ Y      │

In [112]:
pool!(df, [:A, :B])

UndefVarError: UndefVarError: pool! not defined

# Web scraping

Real-world use cases also include scraping data from the Web for analysis. Let's build a small web scraper to fetch Reddit posts.


For this, we will need the JSON and Requests packages:


In [126]:
Pkg.add("JSON")

  Resolving package versions...
   Updating `~/.julia/environments/v1.4/Project.toml`
 [no changes]
   Updating `~/.julia/environments/v1.4/Manifest.toml`
 [no changes]


In [133]:
Pkg.add("HTTP")

  Resolving package versions...
   Updating `~/.julia/environments/v1.4/Project.toml`
  [cd3eb016] + HTTP v0.8.16
   Updating `~/.julia/environments/v1.4/Manifest.toml`
 [no changes]


In [134]:
using JSON, HTTP

┌ Info: Precompiling HTTP [cd3eb016-35fb-5094-929b-558a96fad6f3]
└ @ Base loading.jl:1260


In [171]:
response = HTTP.get("$(reddit_url)/.json")

HTTP.Messages.Response:
"""
HTTP/1.1 200 OK
Connection: keep-alive
Content-Length: 113566
Content-Type: application/json; charset=UTF-8
x-ua-compatible: IE=edge
x-frame-options: SAMEORIGIN
x-content-type-options: nosniff
x-xss-protection: 1; mode=block
access-control-allow-origin: *
access-control-expose-headers: X-Moose
cache-control: max-age=0, must-revalidate
X-Moose: majestic
Accept-Ranges: bytes
Date: Thu, 23 Jul 2020 07:39:13 GMT
Via: 1.1 varnish
X-Served-By: cache-bom4744-BOM
X-Cache: MISS
X-Cache-Hits: 0
X-Timer: S1595489952.433088,VS0,VE612
Vary: accept-encoding
Set-Cookie: loid=00000000007dpmnkze.2.1595489952545.Z0FBQUFBQmZHVDZnQjd2X2h5QllTVWliSnFFWmREX1J1a3AzdlBlbzJHUFRGZ0NiUXJHRklORnB1WlJpOFJXOERuSVVYTkhYaTVoRHNCWjFHd3pBNEktUGJsNXo1MzZFbC1SWUlla3gyY0dudW9famtldnhiUFV1UlVVaGNNemR2UElrQnJFNnRJRmI; Domain=reddit.com; Max-Age=63071999; Path=/; expires=Sat, 23-Jul-2022 07:39:12 GMT; secure; SameSite=None; Secure
Set-Cookie: session_tracker=MiAuj6sjiqNg5864vB.0.1595489952545.Z0FB